### ISSUE TITLE
Flat rendering of inline style ranges with complex inline styles

### ISSUE THREAD

**Issue by acusti** [lokiuz/redraft#23]
Thanks so much for building this! We’ve implemented it in our app for rendering draft.js text content when that content is not in text editing mode, and when there are many separate pieces of text content on a page at once, it has been a considerable perf improvement (we serialize and store text content as the result of `convertToRaw`, so that’s how it lives in our Redux store). We are also using it for generating HTML for that content server-side, which proved necessary for us because we wanted to use `<p>` instead of `<div>` as our default block element, but draft.js renders `<div>s` inside the `<p>` block element, which is invalid HTML and which React’s `renderToString` function actually doesn’t allow (it creates an empty `<p>` element and takes the `<div>` with its children that was inside it and puts it after the empty `<p>`). Using redraft, we’ve been able to simplify and cleanup the output when not in text editing mode and render valid, semantic HTML on the server, so thank you very much for making it!

As described in the draft.js docs on [complex inline styles](https://facebook.github.io/draft-js/docs/advanced-topics-inline-styles.html), in editorState, draft.js stores inline styling on a character-by-character level, meaning that nested styles (bold text inside an italicized range, for example) are actually represented as a flat set of `['BOLD', 'ITALIC']`; this means that when you use the `customStyleMap` to render those inline styles, draft converts the nested range into

```
<span style={ { fontWeight: 'bold', fontStyle: 'italic' } }>{ text }</span>
```
, as opposed to

```
<span style={ { fontStyle: 'italic' } }>`<span style={ { fontWeight: 'bold'  } }>{ text }</span></span>
```

The depth of inline element rendering never grows past 1. This doesn’t seem to be the case when using redraft, I’m assuming because `convertToRaw` takes the character-by-character OrderedSet and converts it to an array of inline style ranges. Can you think of a way I could use the current redraft API to ensure that styled inline elements never nest? If not, would you be willing to explore expanding or modifying the API to support a mode in which inline style ranges get rendered the way they are rendered in draft.js?

Thanks again!

**Comment by lokiuz** [comment]
Hey, I'm very glad you like it.

When I was building redraft, I was aiming to be able to render semantically correct output from draft raw. That means that rather than `<span style={ { fontStyle: 'italic' } }>` it would render `<em>` and that also includes nesting those tags.

I might look into allowing redraft to support an api which is more similar to what draft.js uses as per #20 so I'll keep this in mind as well. Not sure when though so any PRs are also welcome. 😄 

It also might be possible to hack a flatter render using [react children API](https://facebook.github.io/react/docs/react-api.html#react.children.only), but it will probably only work if the style ranges align.
Check the child type and if its same type and the only child merge those styles and only render the child's children to remove the nested wrapper. Finally use the same component in redraft to render different inline styles passing props for each style. (Did that make any sense?)

**Comment by acusti** [comment]
Thanks for the suggestion! I 100% understand not getting to PRs promptly, as I’m only responding 3 weeks after your comment. I’m going to explore hacking together a flatter render for cases where style ranges align, because in practical terms, that seems to be the case quite often for our users, so should help clean up the markup considerably in a lot of cases. I’ll post a code snippet here when I have something that works well.

**Comment by acusti** [comment]
Wasn’t too tricky after all. This code snippet assumes you have a `customStyleMap` ([API reference](https://draftjs.org/docs/api-reference-editor.html#customstylemap)) object.

```js
const renderInlineStyle = (style, children, { key }) => {
    let elementChild;
    const hasSingleElementChild = children.every((child) => {
        if (React.isValidElement(child) && !elementChild) {
            elementChild = child;
            return true;
        }
        return false;
    });

    if (hasSingleElementChild && elementChild) {
        return React.cloneElement(elementChild, {
            style: { ...elementChild.props.style, ...style },
        });
    }

    return (
        <span key={ key } style={ style }>
            { children }
        </span>
    );
};

const renderers = {
    /**
     * These callbacks will be called recursively to render a nested structure
     */
    inline: _.map(customStyleMap, (style, key) => renderInlineStyle.bind(null, style)),
    ...
};

const Renderer = (props) => (
    <div>{ redraft(props.raw, renderers) }</div>
);
```

**Comment by lokiuz** [comment]
Just pushed a new beta version to npm with a new api for a flatter render. If you don't mind some testing and feedback would be awesome! A simple how-to is included in the updated [readme](https://github.com/lokiuz/redraft/blob/next/README.md#using-style-renderer-instead-of-inline). You can install using the next tag ie. `npm install redraft@next`

**Comment by acusti** [comment]
I’ll definitely try it out! It’ll take a bit of time before I can get to it, but shouldn’t be too long.

If you want to see our app to see where we are redraft + draft, sign up at https://app.brandcast.io/signup, then go through the first part of the get started flow (the 2nd lesson introduces text editing).

**Comment by lokiuz** [comment]
Closing this, as I went ahead and published 0.8.0

**Comment by acusti** [comment]
Sorry I didn’t follow up back in March and April,, but we’ve been using the new flatter rendering in 0.8.0 for a couple months now without issue. Thanks again!
